### PR TITLE
Issue #3621: Provide custom exception method to indicate error to AJAX.

### PR DESCRIPTION
**Pull request by stefanhaerter**
Closing #3621

### CODE REVIEW

File: Kernel/Output/HTML/Layout.pm [hunk 1]
```diff
@@ -2919,6 +2920,50 @@ sub JSONReply {
...
+    my ( $Self, %Param ) = @_;
+
+    # if given, check validity of status code
+    # allowed are client errors (400-499) and server errors (500-599)
+    my $StatusCode;
+    if ( $Param{StatusCode} ) {
+        if ( !is_client_error( $Param{StatusCode} ) && !is_server_error( $Param{StatusCode} ) ) {
+            $Kernel::OM->Get('Kernel::System::Log')->Log(
+                Priority => 'error',
+                Message  => "Status code $Param{StatusCode} is not a valid HTTP status!",
```
Comment (bschmalhofer) [inline Kernel/Output/HTML/Layout.pm]: The message could be more specific "is not an error HTTP status"

**Review by bschmalhofer — APPROVED** [review]
Looks good to me

File: Kernel/Output/HTML/Layout.pm (hunk 1 above)
Comment (stefanhaerter) [reply Kernel/Output/HTML/Layout.pm]: Thank you :) Changed this.

File: Kernel/Output/HTML/Layout.pm [hunk 2]
```diff
@@ -2919,6 +2920,50 @@ sub JSONReply {
...
+        if ( !is_client_error( $Param{StatusCode} ) && !is_server_error( $Param{StatusCode} ) ) {
+            $Kernel::OM->Get('Kernel::System::Log')->Log(
+                Priority => 'error',
+                Message  => "Status code $Param{StatusCode} is not a valid HTTP error status!",
+            );
+            return;
+        }
+        $StatusCode = $Param{StatusCode};
+    }
+    $StatusCode ||= 500;
```
Comment (bschmalhofer) [inline Kernel/Output/HTML/Layout.pm]: Being a bit nitpickey here, I suggest to write the assignment as
`my $StatusCode = $Param{StatusCode} || 500;`
The reasoning is that variables should have limited scope.

Comment (stefanhaerter) [reply Kernel/Output/HTML/Layout.pm:2952]: But how do you then access the variable for usage in `Plack::Response->new()`, if you scope it to the if-Block?

Comment (stefanhaerter) [reply Kernel/Output/HTML/Layout.pm]: What do you think about the updated version I just pushed?

Comment (bschmalhofer) [reply Kernel/Output/HTML/Layout.pm]: Excellent, that is better than mine. First handle the default case and then the sanity check.

**Pull request merged**